1-Draft/RFCNNNN-Download-Cmdlet.md

```diff
@@ -0,0 +1,294 @@
+The ability to download files over multiple protocols is an essential feature.
+A PowerShell user will want a PowerShell native way to perform this task.
+
+As there is a feature gap that is essential to fill
+and the current Web Cmdlets are not an ideal candidate for closing that gap,
+a new cmdlet should be added whose primary objective is to support file downloads over multiple protocols.
+
+## Specification
+
+`Invoke-Download` shall be the name of the cmdlet with the alias of `idl`.
+
+### Syntax
+
+The syntax shall be as follows:
+
+```none
+Invoke-Download -Uri <Uri> [-Credential <PSCredential>]
+    [-Path <String> | -LiteralPath <String>]
+    [-Proxy <Uri>] [-ProxyCredential <PSCredential>]
+    [-Resume] [-PassThru] [-AsJob]
+```
+
+```none
+idl -Uri <Uri> [-Credential <PSCredential>]
+    [-Path <String> | -LiteralPath <String>]
+    [-Proxy <Uri>] [-ProxyCredential <PSCredential>]
+    [-Resume] [-PassThru] [-AsJob]
+```
+
+Examples:
+
+```powershell
+idl https://i.imgur.com/gXnRf4J.jpg
+Invoke-Download -Uri 'https://i.imgur.com/gXnRf4J.jpg'
+Invoke-Download -Uri 'https://i.imgur.com/gXnRf4J.jpg' -Path 'c:\images\rin.jpg'
+Invoke-Download -Uri 'https://i.imgur.com/gXnRf4J.jpg' -PassThru | Get-content
+```
+
+### Relationship with Web Cmdlets
+
+The cmdlet shall not share base code with the existing Web Cmdlets.
+The cmdlet shall have a subset of the Web Cmdlets features that are specific or essential to downloading files.
+For example, the cmdlet will not support body content nor HTTP methods other than GET.
+
+As the cmdlet is not centered HTTP and HTTPS,
+the available features need to be generalized for multiple protocols.
+To this end, advanced authentication features in the Web Cmdlets will not exist in the download cmdlet.
+Most protocols expect a username and password.
+At minimum, a `PSCredential` will be accepted.
+
+### Automatic Filename Resolution Feature
+
+A key feature of the cmdlet is to require nothing more than a URI.
+Commands such as `curl` and `wget` have long supported automatic file name resolution.
+This allows the user to supply just a URI and the file will be saved in the current working directory.
+The file name is automatically chosen based on the URI and the presence of existing files.
+
+The cmdlet will automatically chose a file name when a path is not supplied.
+The filename will be chosen in this manner:
+
+1. The last segment of `Uri.Segments`
+1. In the case that the last segment is `/`, the file name `index.html` will be used
+1. Illegal file name characters will be replaced with their URL encoded equivalent.
+1. If a file with that name is present, the basename will be appended with a period followed by `DateTime.Now.ToString("yyyMMddHHmmssfff")`
+1. The above rule will repeat until it locates a name that does not exist or 100 attempts
+1. If in the unlikely event all the above fails, an error will be generated and the user will be suggested to supply a path or change directories.
```

Comment:
Worth clarifying here - and in all other cases below - whether the error is _non_-terminating or _statement_-terminating.

Reply:
It's intentionally not clarified. I have not yet decided if it is worth expanding this to handle multiple URIs. if it is extended to handle multiple URIs it would not be terminating. If it is not extended.. it would end the cmdlet anyway as there is nothing left to process. 

In other words.. this is up for discussion.